### ISSUE TITLE
product_ondemand view fails to be created due to missing column 'parent_item.pick_reason_code'

### ISSUE THREAD

**Issue by djicode** [openboxes/openboxes#1390]
When setting up a new instance and database migrations are performed sequentially:
changelog-2019-01-26-0842-create-view-demand.xml
fails on 
Caused by: com.mysql.jdbc.exceptions.jdbc4.MySQLSyntaxErrorException: Unknown column 'parent_item.pick_reason_code' in 'field list'

Adding the view manually after migrations have completed and 'parent_item.pick_reason_code' exists fixes the issue.

**Comment by jmiranda** [comment]
What version of OpenBoxes are you trying to deploy? More specifically, where did you get the WAR file? From the Releases page? Or did you build it yourself off master/develop?

**Comment by djicode** [comment]
built off the develop branch myself.

**Comment by djicode** [comment]
Testing with Release 0.8.10 now.

**Comment by jmiranda** [comment]
Ok thanks, I'll take a look at develop. It's odd that I hadn't seen this issue yet.

**Comment by djicode** [comment]
No issues with the 0.8.10 release version.

**Comment by jmiranda** [comment]
Thanks for reporting back. We're working on finalizing 0.8.11 so I'll debug this issue before that release goes out the door.

**Comment by jmiranda** [comment]
TL;DR: I need to review this more when I get a chance, but you should be ok going forward. Basically, this error goes away, once the server starts up and the Liquibase migrations have been executed.

I just realized that this occurs during start up when the Liquibase migrations are being run. The column in question has not been created yet. However, there are always background processes (Quartz jobs) that are running to populate reporting / data warehouse tables. In this case, the background process is executed before the column has been created. Eventually the startup process will complete, the column will be created, and the background processes will executed without failing. Unfortunately, there's no built-in way to delay the background processes, so we just need to live with these errors in the log until the server has finished the startup process. 

We solved this way back in 0.8.0 https://github.com/openboxes/openboxes/issues/26, by delaying the start of the database-dependent background processes until Liquibase migrations were complete. However, I ran into some weird issues with threads getting stuck so I've been reluctant to apply the same solution to all Quartz jobs.

**Comment by djicode** [comment]
Makes sense. I was wondering why you were postponing some scheduled tasks but not the database ones that were failing to populate the unmigrated tables. I figured it had to do with some issue with the liquibase migrations. I've never used liquibase so was not sure if it could reconcile the failed migration of 'changelog-2019-01-26-0842-create-view-demand.xml' and would subsequently retry later.

No worries. Thanks!

________________________________

**Comment by jmiranda** [comment]
Actually, I'm wrong here. As you showed in the log snippet, the error is occurring on the changeset, not from within one of the background jobs. So I think what's actually happening is that the ordering of the changesets is causing this error. The product demand view was created back in January of last year and we use the runOnChange attribute of the changeSet and replaceIfExists of the createView operation to do just that ... replace the view definition if the view definition has changed. However, the problem is that no matter when the view is changed this changeset is executed BEFORE some of the other changesets it might depend on. See below. Event though we edited the view definition in October, it still gets executed before the reason code column is added.

```
2020-01-13 16:51:20,823 [localhost-startStop-1] INFO  liquibase  - Change set 0.8.x/changelog-2019-01-26-0842-create-view-demand.xml::1546667174939-1::adminuser (generated) failed, but failOnError was false
liquibase.exception.JDBCException: Error executing SQL CREATE OR REPLACE VIEW `openboxes`.`product_demand` AS SELECT

...

2020-01-13 16:52:07,434 [localhost-startStop-1] INFO  liquibase  - Changeset 0.8.x/changelog-2019-08-20-1521-add-column-requisition-item-pick-reason-code.xml::1566332763499-1::jmiranda (generated)::(MD5Sum: 18a1b067e673cc36ef3ed663377a141)

...

2020-01-13 16:52:11,998 [localhost-startStop-1] INFO  liquibase  - Release Database Lock
2020-01-13 16:52:12,001 [localhost-startStop-1] INFO  liquibase  - Successfully released change log lock
2020-01-13 16:52:12,182 [localhost-startStop-1] INFO  bootstrap.BootStrap  - Finished running liquibase changelog(s)!
```

**Comment by jmiranda** [comment]
In other words, this needs to be dealt with. I probably just need to create a new changeset every time I edit the view.
